content/blog/2022/2022-08-03-feature-flags-in-nodejs-react-client-with-launchdarkly.md

```diff
@@ -0,0 +1,305 @@
+
+To specify which variation a feature flag will show to its user, we can set **targeting rules**. A targeting rule for a feature flag is inactive by default. The most straightforward targeting criterion is "display variation X to all users." Show variation A for all users with attribute X, variation B for all users with attribute Y, and variation C for all other users is a more intricate targeting rule.
+
+Instead of a polling design, LaunchDarkly makes use of a [streaming architecture](https://launchdarkly.com/blog/launchdarklys-evolution-from-polling-to-streaming/). From a scalability point of view, this architecture is beneficial so that our application doesn't have to make a network call each time we need to analyze or fetch a feature flag. Furthermore, feature flag evaluation will continue to function even if the LaunchDarkly server has stopped responding to our calls, which is beneficial for resilience.
+
+## Create a Simple React App
+
+Let’s start by creating a simple React UI application. We will build this using one of the most popular npm libraries, [create-react-app](https://create-react-app.dev/). It is one of the simplest methods to start a new React project, making it a great option for both serious, large-scale apps and your own, individual projects.
+
+We will use the tool npx which gives us the ability to use the `create-react-app` package without having to first install it on our computer, which is very convenient. It also ensures that we are always using the latest version of `create-react-app`.
+
+```bash
+npx create-react-app nodejs-react-feature-flag-launchdarkly
+```
+
+This will create a folder with the name “nodejs-react-feature-flag-launchdarkly” and automatically install all the required packages inside that folder. The folder tree would look something like below:
+
+```bash
+nodejs-react-feature-flag-launchdarkly
+├── README.md
+├── node_modules
+├── package.json
+├── .gitignore
+├── public
+└── src
+```
+
+Now we can simply run our app and see a dev server starting up and opening a UI with the React logo:
+
+```bash
+npm start
+```
+
+The development server will start up at http://localhost:3000 and we can see the home page.
+
+{{% image alt="React UI" src="images/posts/nodejs-frontend-launchdarkly/React_Logo.png" %}}
+
+Next we will add the [launchdarkly-js-client-sdk](https://github.com/launchdarkly/js-client-sdk) package to use client-side methods to fetch the feature flag variations:
+
+```bash
+npm install launchdarkly-js-client-sdk
+```
+
+
+
+## String-based Feature Flag to Hide/Display Content on the Fly
+
+In our first use-case, let’s try to simply define a string feature flag in LaunchDarkly UI. We will simply list some users in the React UI. If any of the users matches the user name defined in the feature flag variation, then we will encrypt that user name in the UI. It would be pretty simple but enough to understand how we can enable/disable a feature in web UI.
```

Comment:
The use case of "encrypting" a certain user name in the UI with a feature flag is not really compelling, I'm afraid. Why would anyone do this? Also, it mixes feature flag variations with feature flag targeting (because we're using a user name as a variation, when user names should be used in targeting rules). 

Maybe create a section in the UI that says "shiny new feature" and then show/hide it based on the feature flag. Then, show that we can create targeting rules for different users and if we use one user name, it shows, and if we use another user name, it doesn't show. Also explain that the user name will come from the user's session in real life, instead of being hard-coded as in the code example.

Reply:
I have modified this sub-topic and renamed it "Enable/Disable a Feature on the Fly". I have created a button and hiding it based upon the name of the user string match.